### ISSUE TITLE
Signing algorithm not listed when it could? (and would bring a valuable information)

### ISSUE THREAD

**Issue by jarthod** [rbsec/sslscan#264]
Hi, I am currently trying to debug some problems related to servers wrong signing algorithm ([more details here](https://updown.uservoice.com/knowledgebase/articles/1989757)) and tried to use `sslscan 2.0.7` for that (thanks for this awesome tool BTW :bow:)

I am comparing two servers:
1. First one is apparently only supporting "RSA-SHA1" signing algorithm (which is deprecated and cause errors with recent versions of openssl), I don't want to share the hostname publicly because it's one of my clients but I'll share the results of the commands (I can share it privately if needed), let's say this is `$HOSTNAME` in the commands below.
2. Second one is my own domain for comparison (updown.io) which support multiple algorithms and does not cause any trouble.

I noticed with some manual testing using `openssl 1.1.1` (older version which does not reject the RSA-SHA1 algorithm yet) that this first server always respond with SHA1 no matter what the client asks:

```
> echo | openssl s_client -servername $HOSTNAME -connect $HOSTNAME:443 -sigalgs "RSA-PSS+SHA256" | fgrep "Peer sign"
[...]
Peer signing digest: SHA1
Peer signature type: RSA
> echo | openssl s_client -servername $HOSTNAME -connect $HOSTNAME:443 -sigalgs "RSA+SHA224" | fgrep "Peer sign"
[...]
Peer signing digest: SHA1
Peer signature type: RSA
```
(same thing with any algorithm), and so this website fails with more recent version of OpenSSL because [SHA1 is deprecated](https://datatracker.ietf.org/doc/html/draft-ietf-tls-md5-sha1-deprecate).

On the other hand with a working website like updown.io, we can see many algorithms are supported (even the deprecated `RSA-SHA1`) but this works fine because the client can use a better algorithm (`RSA-PSS+SHA256` by default):

```
> echo | openssl s_client -servername updown.io -connect updown.io:443 -tls1_2 -sigalgs "RSA-PSS+SHA256" | fgrep "Peer sign"
[...]
Peer signing digest: SHA256
Peer signature type: RSA-PSS
> echo | openssl s_client -servername updown.io -connect updown.io:443 -tls1_2 -sigalgs "RSA+SHA1" | fgrep "Peer sign"
[...]
Peer signing digest: SHA1
Peer signature type: RSA
```

So after seing that I thought: maybe I can use `sslscan` to more easily show available algorithms and make it easier for my clients to check their own servers (expecting to see here that only RSA-SHA1 is supported).

But when I tried I saw this:
```
> sslscan --show-sigs $HOSTNAME
Version: 2.0.7
OpenSSL 1.1.1l  24 Aug 2021
[...]
  Server Signature Algorithm(s):
TLSv1.2  Server accepts all signature algorithms.
[...]
```

Whait, what? surely it does not :thinking: 
On updown.io the results looks acurate:

```
> sslscan --show-sigs updown.io
Version: 2.0.7
OpenSSL 1.1.1l  24 Aug 2021
[...]
  Server Signature Algorithm(s):
TLSv1.3  rsa_pss_rsae_sha256
TLSv1.3  rsa_pss_rsae_sha384
TLSv1.3  rsa_pss_rsae_sha512
TLSv1.2  rsa_pkcs1_sha1
TLSv1.2  rsa_pkcs1_sha224
TLSv1.2  rsa_pkcs1_sha256
TLSv1.2  rsa_pkcs1_sha384
TLSv1.2  rsa_pkcs1_sha512
TLSv1.2  rsa_pss_rsae_sha256
TLSv1.2  rsa_pss_rsae_sha384
TLSv1.2  rsa_pss_rsae_sha512
[...]
```

So I had a quick look at the code (c95e9fba82ee04ef2505ecbb1cd8cd092e108963), and it looks like this message is actually designed for this use-case (server accepts anything). But wouldn't it be better to check what the server actually sends back? (like what I see in the output of the `openssl` command) to compute the actual list of what is supported server-side and not just what say that everything is silently accepted without compying? Alternatively maybe we could improve this message to at least display what the server uses during the bogus algorithm test?

On the same topic I looked at other issues and noticed #207 where for google.com, `sslscan` also displays:
```
> sslscan --show-sigs google.com
Version: 2.0.7
OpenSSL 1.1.1l  24 Aug 2021
[...]
Server Signature Algorithm(s):
TLSv1.3  Server accepts all signature algorithms.
[...]
```

So the same output, but this site is working better, and actually it's not that buggy, if we give it different (valid) algorithms it uses them:
```
> echo | openssl s_client -servername google.com -connect google.com:443 -sigalgs "RSA-PSS+SHA256" | fgrep "Peer sign"
[...]
Peer signing digest: SHA256
Peer signature type: RSA-PSS
> echo | openssl s_client -servername google.com -connect google.com:443 -sigalgs "RSA-PSS+SHA512" | fgrep "Peer sign"
[...]
Peer signing digest: SHA512
Peer signature type: RSA-PSS
```

So in this case maybe they are responding to bogus algorithm requests with another one for compatibility reasons, but they still honor requests for actual algorithm and so we could make a list too. And surely they do not use `RSA-SHA1` here. The most important being that in the end, both hostname shows exactly the same output despite having very different behavior (one of them being impossible to handshake with a recent OpenSSL client so quite an important difference).

ps: my C is pretty rusty but I might try to draft a PR if you want, I would like your opinion first though before I spend too much time on this :)
ps2: are you accepting financial contributions somewhere? I try to give back more and more to the many open source projects I use, be it in the form of code contributions or tips.

**Comment by jtesta** [comment]
Hi Adrien.  I wrote the signature enumeration code for sslscan, so I can help.

Back when I wrote the code, I noticed that some servers will accept all signature algorithms presented in the handshake.  To compensate, I made sslscan first submit a bogus signature algorithm (see https://github.com/rbsec/sslscan/blob/master/sslscan.c#L5790) and check to see if the server accepts it.  If so, sslscan concludes that the server will accept anything and won't continue testing.  Without this check, sslscan will report all server signatures back to the user.  While this is technically accurate, I thought it would be better to just tell the user that everything is being accepted.

To see this in action, you can comment out line [5792](https://github.com/rbsec/sslscan/blob/master/sslscan.c#L5792), recompile, and run against google.com:

```
$ ./sslscan google.com
Version: 2.0.13-wip-static
OpenSSL 1.1.1o-dev  xx XXX xxxx

Connected to 142.250.80.46

Testing SSL server google.com on port 443 using SNI name google.com

[...]

  Server Signature Algorithm(s):
TLSv1.3  rsa_pkcs1_nohash
TLSv1.3  dsa_nohash
TLSv1.3  ecdsa_nohash
TLSv1.3  rsa_pkcs1_md5
TLSv1.3  dsa_md5
TLSv1.3  ecdsa_md5
TLSv1.3  rsa_pkcs1_sha1
TLSv1.3  dsa_sha1
TLSv1.3  ecdsa_sha1
TLSv1.3  rsa_pkcs1_sha224
TLSv1.3  dsa_sha224
TLSv1.3  ecdsa_sha224
TLSv1.3  rsa_pkcs1_sha256
TLSv1.3  dsa_sha256
TLSv1.3  ecdsa_secp256r1_sha256
TLSv1.3  rsa_pkcs1_sha384
TLSv1.3  dsa_sha384
TLSv1.3  ecdsa_secp384r1_sha384
TLSv1.3  rsa_pkcs1_sha512
TLSv1.3  dsa_sha512
TLSv1.3  ecdsa_secp521r1_sha512
TLSv1.3  rsa_pss_rsae_sha256
TLSv1.3  rsa_pss_rsae_sha384
TLSv1.3  rsa_pss_rsae_sha512
TLSv1.3  ed25519
TLSv1.3  ed448
TLSv1.3  rsa_pss_pss_sha256
TLSv1.3  rsa_pss_pss_sha384
TLSv1.3  rsa_pss_pss_sha512
TLSv1.2  rsa_pkcs1_nohash
TLSv1.2  dsa_nohash
TLSv1.2  ecdsa_nohash
TLSv1.2  rsa_pkcs1_md5
TLSv1.2  dsa_md5
TLSv1.2  ecdsa_md5
TLSv1.2  rsa_pkcs1_sha1
TLSv1.2  dsa_sha1
TLSv1.2  ecdsa_sha1
TLSv1.2  rsa_pkcs1_sha224
TLSv1.2  dsa_sha224
TLSv1.2  ecdsa_sha224
TLSv1.2  rsa_pkcs1_sha256
TLSv1.2  dsa_sha256
TLSv1.2  ecdsa_secp256r1_sha256
TLSv1.2  rsa_pkcs1_sha384
TLSv1.2  dsa_sha384
TLSv1.2  ecdsa_secp384r1_sha384
TLSv1.2  rsa_pkcs1_sha512
TLSv1.2  dsa_sha512
TLSv1.2  ecdsa_secp521r1_sha512
TLSv1.2  rsa_pss_rsae_sha256
TLSv1.2  rsa_pss_rsae_sha384
TLSv1.2  rsa_pss_rsae_sha512
TLSv1.2  ed25519
TLSv1.2  ed448
TLSv1.2  rsa_pss_pss_sha256
TLSv1.2  rsa_pss_pss_sha384
TLSv1.2  rsa_pss_pss_sha512
```

Notice that the full list of signature algorithms from lines 5793-5821 are returned for both TLSv1.3 and v1.2.

So the code does appear to be functioning properly.  Does this address your original concern?  Otherwise, I'm not sure what you would like sslscan to do instead in this situation.

**Comment by jarthod** [comment]
@jtesta thanks for the quick reply and the explanation!

This is pretty much what I understood from the source code, and I understand showing all tested algorithms instead of just this warning message is pointless. It does not address my original concern though, let me explain:

What would be interesting here IMO is showing what the server actually supports (the algorithm the server used when replying), this is what `sslscan` should do instead in this situation. I am not saying this is easy to do though, I don't know the handhshake protocol well, but as I said in my analysis I see the actually used algorithm in `openssl` output so I suppose we can get it at some point? (it might require going further in the handshake).

So as per my first example if the server responds to every attempt with `RSA-SHA1`, `sslscan` would dedup of course (no need to display `TLSv1.2  rsa_pkcs1_sha1` line 50 times) and show a single line: `TLSv1.2  rsa_pkcs1_sha1` (we can also keep the warning saying that the "Server accepts all signature algorithms"). And in the case of google.com, it would show the couple supported algorithm, and the "default" one they are returning to bogus requests. If we want to be extra precise we could indicate at the corresponding line that this algorithm was used even when asked for other ones (so we know which one it is when some others are working like on google.com).

So basically what I'm suggesting is to display the actually used algorithm instead of the requests one, by getting this from the server response (skipping all duplicates of course), and when they don't match (requested vs actually used), display this warning or another one so people know the implementation is not respecting what is asked.

For normally behaving server this would not change anything in the output.

Do you think this would make sense?

**Comment by jtesta** [comment]
> What would be interesting here IMO is showing what the server
actually supports (the algorithm the server used when replying), this
is what sslscan should do instead in this situation.

That was the goal, but I don't know of any way to do that other than
how its currently implemented.  That is, the client sends a Client
Hello message with a signature_algorithms extension with a single
algorithm specified.  If the server responds with anything other than a
Server Hello, then this is interpreted as not supporting that specific
signature algorithm.  The client attempts all signature algorithms one
at a time in this manner, and outputs the positive results (i.e.: tests
that resulted in a Server Hello response).

If you knew of a more efficient way of enumerating the server's
signature algorithms, however, please let me know!


> So as per my first example if the server responds to every attempt
> with RSA-SHA1, 

In the case where sslscan reports that all signature algorithms are
accepted, what would happen is the server would return a Server Hello
regardless of what we put in the signature_algorithms extension field.
 Hence, sslscan first tries a bogus signature algorithm (as I mentioned
in my previous response) to see if the server is at least testable.

As far as I can see, the server never explicitly responds with anything
it supports.  Then again, TLS responses can be rather complex, so there
could be something buried in there.  If you stare at Wireshark output
for long enough and find something, though, I'd definitely like to hear
about it!

At this point, I'm not seeing a better way to do this testing, nor is
there any extra information that we get that isn't already being
reported.

**Comment by jarthod** [comment]
> As far as I can see, the server never explicitly responds with anything
it supports.  Then again, TLS responses can be rather complex, so there
could be something buried in there.  If you stare at Wireshark output
for long enough and find something, though, I'd definitely like to hear
about it!


Well I don't know the technical details but in my examples using the `openssl` command we can clearly see the algorithm actually used by the server being different from the one asked, example:
```
> echo | openssl s_client -servername $HOSTNAME -connect $HOSTNAME:443 -sigalgs "RSA-PSS+SHA256" | fgrep "Peer sign"
[...]
Peer signing digest: SHA1
Peer signature type: RSA
```

So it must be there somewhere, I suppose I can try spending some time looking for it.
But if I can find this information, do you think it would be a good idea to implement the change I suggest (display the actual algorithm used by the server instead of the one the client asked), we would still have to do N tests for each algorithms.

**Comment by jarthod** [comment]
So I had a look and found the information (the signature algorithm actually used by the server) to be in the "Server Key Exchange" message, after the "Server Hello" and "Certificate" messages, we can see it here:
![image](https://user-images.githubusercontent.com/201687/163436713-8d7e6c21-846b-43e7-9808-6f08008f9fe7.png)

From what I can see in the code we're never going this far in the handshake parsing (we're only parsing "Server Hello"?) so I suppose it'll be a significant amount of work to parse up to here?

**Comment by jtesta** [comment]
Yeah, that would be about 3-4 hours of work with testing included.  The
information we get back for that amount of work seems pretty minimal
(to me, anyway).  So that wouldn't be so much a priority for me to
tackle.

But hey... its open source, so if the maintainer is OK with it... feel
free to give it a shot!

**Comment by jarthod** [comment]
Ok thank you for the confirmation, for me with my limited knowledge of TLS, of this codebase and being rusty in C, it would probably be quite more ^^. I agree the information we get is minimal (it's important only for the 0.1% of servers which are misresponding basically), and I am not using this tool regularly so not worth spending too much of my time on it either at the moment.

I note that @rbsec seems to be OK with it as he :+1: the previous comment. So later if this problem becomes more widespread (supposedly it won't as people upgrade their server stack) or if I start using `sslscan` more, maybe I'll revisit. And if other users are having the same thought they'll have this discussion to start from.

So let's leave it there for now, we can close the issue (and maybe flag it as "not fixed") if you want.

**Comment by rbsec** [comment]
Hi both,

Interesting to see the discussion here - I agree that this could potentially have some value as a check, but I'm afraid that I don't really have much time to spend on sslscan at the moment, so like with jtesta this would be quite lot down my priority list.

TBH, I've never been a huge fan of this check because it's not really actionable. Very few systems let people configure the signature algorithms that are used - so it's usually a finding that they can't really do anything about, which is always annoying.

**Comment by jarthod** [comment]
Thanks for your input @rbsec, I totally understand this is zero priority and I am not requesting to get it done (If I really need this feature I can contribute my time to do it). Also agree this is not very actionable, from what I saw in my example here the problem of honoring what the client asks should be actionnable by upgrading outdated software, but this can vary from easy-peasy (there's a new version you just need to install) to totally impossible (no new version or last version does not fix and you can't easily change software). But as we said already this is a small case, we have other ways to show/verify this problem (using `openssl`) and people will naturally upgrade over time so we should have this situation less and less often hopefully.
